### DISCUSSION TITLE
Ishell issues in threading

### DISCUSSION THREAD

**Discussion by dhutter** [mhammond/pywin32#2545] [General]
Given the following small function

```
 def` testdevices ():
     topshell = shell.SHGetDesktopFolder()
     drives_pidl = shell.SHGetSpecialFolderLocation(0, shellcon.CSIDL_DRIVES)
     ishell = topshell.BindToObject(drives_pidl, None, shell.IID_IShellFolder)
     items = ishell.EnumObjects(None, shellcon.SHCONTF_FOLDERS)
     item = items.Next(1)
     while item != []:
         print (ishell.GetDisplayNameOf(item[0], 1))
         item = items.Next(1)    
```

Calling testdevices() inside the main thread of Python works well and without any problems. However, calling the same function within a threading.thread like
usbwatch = threading.Thread(target=testdevices)
causes an exception:
>  File ***, line ***, in testdevices
>    print (ishell.GetDisplayNameOf(item[0], 1))
>  pywintypes.com_error: (-2147221231, 'ClassFactory kann angeforderte Klasse nicht liefern', None, None)

Changing slightly the above function by requesting shellcon.CSIDL_CONNECTIONS instead of shellcon.CSIDL_DRIVES raises an exception in BindToObject. Again, this error only occurs when called within a threading.Thread

>     ishell = topshell.BindToObject(drives_pidl, None, shell.IID_IShellFolder)
> pywintypes.com_error: (-2147221008, 'CoInitialize wurde nicht aufgerufen.', None, None)

I downloaded pywin32 two months ago; so this should be a reasonably new version.

Thanks for your help (in advance) ;-)

**Comment by mhammond** [comment]
This is a support request, as described in the template you removed when opening this issue. Please see https://github.com/mhammond/pywin32#support

**Comment by mhammond** [comment]
(But you should google about COM's threading model - it has very strict rules - you can't just pass arbitrary COM objects to arbitrary threads.)
